[docs] app commands guide (#9557)
Rapptz/discord.py: docs/guide/interactions/slash_commands.rst

.. image:: /images/guide/app_commands/repeat_command_described.png

In addition to the decorator, parameter descriptions can also be added using
Google, Sphinx or Numpy style docstrings.

Examples using a command to add 2 numbers together:

.. code-block:: python

    @client.tree.command() # numpy
    async def addition(interaction: discord.Interaction, a: int, b: int):
        """adds 2 numbers together.

        Parameters
        -----------
        a: int
            left operand
        b: int
            right operand
        """

        await interaction.response.send_message(f"{a} + {b} is {a + b}!")

    @client.tree.command() # google
    async def addition(interaction: discord.Interaction, a: int, b: int):
        """adds 2 numbers together.

        Args:
            a (int): left operand
            b (int): right operand
        """

    @client.tree.command() # sphinx
    async def addition(interaction: discord.Interaction, a: int, b: int):
        """adds 2 numbers together.

        :param a: left operand
        :param b: right operand
        """

If both are used, :func:`.app_commands.describe` always takes precedence.

Naming
^^^^^^^

> Since parameter names are confined to the rules of Python's syntax,

Reviewer: Slash command parameter also have confined rules, I feel like they should be mentioned in here too.
https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-naming

    @app_commands.rename(n_times="number-of-times")
    async def repeat(interaction: discord.Interaction, content: str, n_times: int):
        to_send = textwrap.shorten(f"{content} " * n_times, width=2000)
        await interaction.response.send_message(to_send)

When referring to a renamed parameter in other decorators, the original parameter name should be used.
For example, to use :func:`~.app_commands.describe` and :func:`~.app_commands.rename` together:

.. code-block:: python

    @client.tree.command()
    @app_commands.describe(
        content="the text to repeat",
        n_times="the number of times to repeat the text"
    )
    @app_commands.rename(n_times="number-of-times")
    async def repeat(interaction: discord.Interaction, content: str, n_times: int):
        to_send = textwrap.shorten(f"{content} " * n_times, width=2000)
        await interaction.response.send_message(to_send)

Choices
++++++++

To provide the user with a list of options to choose from for an argument, the :func:`.app_commands.choices` decorator can be applied.

A user is restricted to selecting a choice and can't type something else.

Each individual choice contains 2 fields:

> - A name, which is what the user sees

Reviewer: ```suggestion
- A name, which will be displayed to the user in the client.
```

> - A value, which is hidden to the user and only visible to the API. Typically, this is either the same as the name or something more developer-friendly. Value types are limited to either a :class:`str`, :class:`int` or :class:`float`.

Reviewer: Not 100% sure about my own suggested wording here, but I think it is important to make clear that it is the actual value that gets given to the command by discord.
```suggestion
- A value, which will be the real value provided to the command callback by Discord. Typically, this is either the same as the name or something more developer-friendly. Value types are limited to either a :class:`str`, :class:`int` or :class:`float`.
```

Reviewer: Yeah, shows how little I actually work with choices vs. auto-completion. 
I incorrectly assumed it was analogous to auto-completion, which gives you only the actual value of the choice. 

I still find the phrasing of "only visible to the API" sub-optimal, since it reads it to me like it is only visible to the discord api, not your bot, at least in my reading of me putting myself in the position of someone new to the concepts of discord.py or slash commands.
Maybe rephrase to "only visible to your bot" or "only visible to your command callback". Even though I am also iffy on that phrasing entirely. 
just throwing out one last idea here:
```
A value, that should represent the actual value we want to use inside our command callback. 
This can either be the same as the name or something "machine readable"
- like an integer value (see example below)
- a column id in the database
- a mapping key for an internal dictionary
- or any other representation we need.
```

        app_commands.Choice(name="Blue", value=0x0000FF)
    ])
    async def colour(interaction: discord.Interaction, colour: app_commands.Choice[int]):
        """show a colour"""

        embed = discord.Embed(title=colour.name, colour=colour.value)
        await interaction.response.send_message(embed=embed)

On the client:

.. image:: /images/guide/app_commands/colour_command_preview.png
    :width: 400

discord.py also supports 2 other pythonic ways of adding choices to a command,
shown :func:`here <discord.app_commands.choices>` in the reference.

Autocompletion
+++++++++++++++

Autocomplete callbacks allow the bot to dynamically return up to 25 choices
to a user as they type a parameter.

In short:

- User starts typing.

- After a brief debounced pause from typing, Discord requests a list of choices from the bot.

- An autocomplete callback is called with the current user input.

- Returned choices are sent back to Discord and shown in the user's client.

  - An empty list can be returned to denote no choices.

Attaching an autocomplete function to a parameter can be done in 2 main ways:

1. From the command, with the :meth:`~.app_commands.Command.autocomplete` decorator

> 2. With a separate decorator, :func:`.app_commands.autocomplete`

Reviewer: It is also possible to attach an autocompletion via a [Transformer](https://discordpy.readthedocs.io/en/latest/interactions/api.html?highlight=transformer%20autocomplete#discord.app_commands.Transformer.autocomplete)

+-----------------+------------------------------------------------------------------------------------+
| ``ROLE``        | :class:`~discord.Role`                                                             |
+-----------------+------------------------------------------------------------------------------------+
| ``MENTIONABLE`` | :class:`~discord.User` or :class:`~discord.Member`, or :class:`~discord.Role`      |
+-----------------+------------------------------------------------------------------------------------+
| ``ATTACHMENT``  | :class:`~discord.Attachment`                                                       |
+-----------------+------------------------------------------------------------------------------------+

:ddocs:`Application command option types <interactions/application-commands#application-command-object-application-command-option-type>` as documented by Discord.

User parameter
^^^^^^^^^^^^^^^

Annotating to either :class:`discord.User` or :class:`discord.Member` both point to a ``USER`` Discord-type.

The actual type given by Discord is dependent on whether the command was invoked in DM-messages or in a guild.

> For example, if a parameter annotates to :class:`~discord.Member`, and the command is invoked in a guild,
> discord.py will raise an error since the actual type given by Discord,
> :class:`~discord.User`, is incompatible with :class:`~discord.Member`.

Reviewer: ```suggestion
For example, if a parameter annotates to :class:`~discord.Member`, and the command is invoked in a DM,
discord.py will raise an error since the actual type given by Discord,
:class:`~discord.User`, is incompatible with :class:`~discord.Member`.
```

        if isinstance(user, discord.Member):
            joined = user.joined_at
            if joined:
                relative = discord.utils.format_dt(joined, "R")
                info = f"{info} (joined this server {relative})"

        await interaction.response.send_message(info)

Command groups
---------------

To make a more organised and complex tree of commands, Discord implements command groups and subcommands.
A group can contain up to 25 subcommands or subgroups, with up to 1 level of nesting supported.

Meaning, a structure like this is possible:

.. code-block::

    todo
    ├── lists
    │   ├── /todo lists create
    │   └── /todo lists switch
    ├── /todo add
    └── /todo delete

Command groups **are not invocable** on their own.

Therefore, instead of creating a command the standard way by decorating an async function,
groups are created by using :class:`.app_commands.Group`.

This class is customisable by subclassing and passing in any relevant fields at inheritance:

.. code-block:: python

    class Todo(app_commands.Group, name="todo", description="manages a todolist"):
        ...

    client.tree.add_command(Todo()) # required!

.. note::

>     Groups need to be added to the command tree manually with :meth:`.CommandTree.add_command`,

Reviewer: What about [GroupCogs](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html?highlight=groupcog#discord.ext.commands.GroupCog)

Also within a cog they are also automatically added 
https://gist.github.com/AbstractUmbra/a9c188797ae194e592efe05fa129c57f#file-07-extension_with_optional_group-py

Or are Cogs out of scope for this guide?

Reviewer: Okay, then I agree with the original phrasing and this feedback can be set to resolved 👍

.. code-block:: python

    @client.tree.command()
    @app_commands.guild_only()
    async def serverinfo(interaction: discord.Interaction):
        assert interaction.guild is not None
        await interaction.response.send_message(interaction.guild.name)

Default permissions
++++++++++++++++++++

This sets the default permissions a user needs in order to be able to see and invoke an app command.

Configured by adding the :func:`.app_commands.default_permissions` decorator when defining an app command:

.. code-block:: python

    @client.tree.command()
    @app_commands.default_permissions(manage_nicknames=True)
    async def nickname(interaction: discord.Interaction, newname: str):
        guild = interaction.guild
        if not guild:
            await interaction.response.send_message("i can't change my name here")
        else:
            await guild.me.edit(nick=newname)
            await interaction.response.send_message(f"hello i am {newname} now")

Commands with this check are still visible and invocable in the bot's direct messages,
regardless of the permissions specified.

To prevent this, :func:`~.app_commands.guild_only` can also be added.

.. warning::

>     This can be overriden to a different set of permissions by server administrators through the "Integrations" tab on the official client,

Reviewer: ```suggestion
    This can be overridden to a different set of permissions by server administrators through the "Integrations" tab on the official client,
```

Custom checks come in two forms:

- A local check, which runs for a single command
- A global check, which runs before all commands, and before any local checks

Global check
^^^^^^^^^^^^^

To define a global check, override :meth:`.CommandTree.interaction_check` in a :class:`~.app_commands.CommandTree` subclass.
This method is called before every command invoke.

For example:

.. code-block:: python

    whitelist = {236802254298939392, 402159684724719617} # cool people only

    class MyCommandTree(app_commands.CommandTree):
        async def interaction_check(self, interaction: discord.Interaction) -> bool:
            return interaction.user.id in whitelist

.. note::

    If your project uses :class:`.ext.commands.Bot` as the client instance,
    the :class:`.CommandTree` class can be configured via
    the ``tree_cls`` keyword argument in the bot constructor:

    .. code-block:: python
        :emphasize-lines: 6

        from discord.ext import commands

        bot = commands.Bot(
            command_prefix="?",
            intents=discord.Intents.default(),
            tree_cls=MyCommandTree
        )

Error handling
---------------

> So far, any exceptions raised within a command callback, any custom checks or in a transformer should just be
> printed out in the program's ``stderr`` or through any custom logging handlers.

Reviewer 2: ```suggestion
By default, exceptions are logged using the library logger if the command does not have any error handlers attached to it.
```
Errors are no longer printed to `stderr` by default since v2.0

Default permissions
++++++++++++++++++++

This sets the default permissions a user needs in order to be able to see and invoke an app command.

Configured by adding the :func:`.app_commands.default_permissions` decorator when defining an app command:

.. code-block:: python

    @client.tree.command()
    @app_commands.default_permissions(manage_nicknames=True)
    async def nickname(interaction: discord.Interaction, newname: str):
        guild = interaction.guild
        if not guild:
            await interaction.response.send_message("i can't change my name here")
        else:
            await guild.me.edit(nick=newname)
            await interaction.response.send_message(f"hello i am {newname} now")

Commands with this check are still visible and invocable in the bot's direct messages,
regardless of the permissions specified.

To prevent this, :func:`~.app_commands.guild_only` can also be added.

.. warning::

    This can be overridden to a different set of permissions by server administrators through the "Integrations" tab on the official client,
    meaning, an invoking user might not actually have the permissions specified in the decorator.

> Custom checks

Reviewer 3: the context menus, checks and cooldown guide is written and waiting on the final review (https://github.com/Rapptz/discord.py/pull/7952/). You could probably just link the checks guide page to this section or wait and leave this section as-is.

Reviewer 3: sounds good!

A response needs to occur within 3 seconds, otherwise this message pops up on Discord in red:

.. image:: /images/guide/app_commands/interaction_failed.png

In practice, it's common to use either of the following two methods:

- :meth:`.InteractionResponse.send_message` to send a message
- :meth:`.InteractionResponse.defer` to defer a response

In the case of deferring, a follow-up message needs to be sent within 15 minutes for app commands.

For example, to send a deferred ephemeral message:

.. code-block:: python

    import asyncio
    import random

    @client.tree.command()
    async def weather(self, interaction: discord.Interaction):
        await interaction.response.defer(ephemeral=True) # indicates the follow-up message will be ephemeral

        weathers = ['sunny', 'clear', 'cloudy', 'rainy', 'stormy', 'snowy']
        await asyncio.sleep(5) # an expensive operation... (no more than 15 minutes!)
        forecast = random.choice(weathers)

        await interaction.followup.send(f'the weather today is {forecast}!')

Syncing
++++++++

In order for this command to show up on Discord, the API needs some information regarding it, namely:

> - The name and description

Reviewer 3: Also, sync when you add or modify permissions (`default_permissions` and `nsfw`). It might be useful to denote when not to sync as well (pretty much if any of the code after the `async def ():` part, and library side checks)

but can be narrowed down to a specific set of channels:

.. code-block:: python

    from typing import Union

    @client.tree.command()
    async def channel_info(interaction: discord.Interaction, channel: discord.abc.GuildChannel):
        # Everything except threads
        pass

    @client.tree.command()
    async def channel_info(interaction: discord.Interaction, channel: discord.TextChannel):
        # Only text channels
        pass

    @client.tree.command()
    async def channel_info(interaction: discord.Interaction, channel: Union[discord.Thread, discord.VoiceChannel]):
        # Threads and voice channels only
        pass

.. note::

    Union types can't mix Discord types.

    Something like ``Union[discord.Member, discord.TextChannel]`` isn't possible.

Refer to the :ref:`type conversion table <type_conversion>` for full information on sub-types.

Describing
+++++++++++

Descriptions are added to parameters using the :func:`.app_commands.describe` decorator,
where each keyword is treated as a parameter name.

.. code-block:: python

    @client.tree.command()
    @app_commands.describe(

>         liquid="what type of liquid is on the wall",

Reviewer 3: Noticed that the param descriptions were not capitalized, since the ones in the examples are. Also missing the command description from what I can tell.

```suggestion
        liquid="What type of liquid is on the wall",
        amount="How much of it is on the wall"
```

Reviewer 3: Yea this makes total sense

        amount="how much of it is on the wall"
    )
    async def bottles(interaction: discord.Interaction, liquid: str, amount: int):
        await interaction.response.send_message(f'{amount} bottles of {liquid} on the wall!')

These show up on Discord just beside the parameter's name:

.. image:: /images/guide/app_commands/bottles_command_described.png

Not specifying a description results with an ellipsis "..." being used instead.

In addition to the decorator, parameter descriptions can also be added using
Google, Sphinx or Numpy style docstrings.

Examples using a command to add 2 numbers together:

> .. code-block:: python

Reviewer 4: this code block could use the inline tabs ext

    @client.tree.command() # numpy
    async def addition(interaction: discord.Interaction, a: int, b: int):
        """adds 2 numbers together.

        Parameters
        -----------

>         a: int

Reviewer 3: Pretty much the same exact thing as the last comment (command and param descriptions not being capitalized). This pretty much applies to the other two docstrings below.

- After a brief debounced pause from typing, Discord requests a list of choices from the bot.

- An autocomplete callback is called with the current user input.

- Returned choices are sent back to Discord and shown in the user's client.

  - An empty list can be returned to denote no choices.

Attaching an autocomplete function to a parameter can be done in 2 main ways:

1. From the command, with the :meth:`~.app_commands.Command.autocomplete` decorator
2. With a separate decorator, :func:`.app_commands.autocomplete`

Code examples for either method can be found in the corresponding reference page.

.. note::

    Unlike :func:`.app_commands.choices`, a user can still submit any value instead of
    being limited to the bot's suggestions.

.. warning::

    Since exceptions raised from within an autocomplete callback are not considered handleable,
    they're silently ignored and discarded.

    Instead, an empty list is returned to the user.

Range
++++++

:class:`str`, :class:`int` and :class:`float` type parameters can optionally set a minimum and maximum value.
For strings, this limits the character count, whereas for numeric types this limits the magnitude.

> To set a range, a parameter should annotate to :class:`.app_commands.Range`.

Reviewer 3: It would be useful to have an example of how you would annotate it

Reviewer 3: Makes sense 👍

.. code-block:: python

    import discord
    from discord import app_commands

    class MyClient(discord.Client):
        def __init__(self):
            super().__init__(intents=discord.Intents.default())
            self.tree = app_commands.CommandTree(self)

    client = MyClient()

.. note::

    If your project instead uses :class:`.ext.commands.Bot` as the client instance,
    a :class:`~discord.app_commands.CommandTree` has already been defined at :attr:`.Bot.tree`,
    so this step is largely skipped.

Creating a command
-------------------

Slash commands are created by decorating an async function.
This function is then called whenever the slash command is invoked.

For example, the following code responds with "meow" on invocation:

.. code-block:: python

    @client.tree.command()
    async def meow(interaction: discord.Interaction):
        """Meow meow meow"""

        await interaction.response.send_message('meow')

Functions of this pattern are called callbacks, since their execution is
left to the library to be called later.

There are two main decorators to use when creating a command:

> 1. :meth:`tree.command() <.CommandTree.command>` (as seen above)
> 2. :func:`.app_commands.command`

Reviewer 4: ```suggestion
1. :meth:`@tree.command() <.CommandTree.command>` (as seen above)
2. :func:`@app_commands.command() <.app_commands.command>`
```
nice to be consistent (also think the @ is a nice touch)

.. image:: /images/guide/app_commands/this_option_is_required.png
    :width: 300

Some parameter types have different modes of input.

For example, annotating to :class:`~discord.User` will show a selection of users to
pick from in the current context and :class:`~discord.Attachment` will show a file-dropbox.

A full overview of supported types can be seen in the :ref:`type conversion table <type_conversion>`.

typing.Optional
++++++++++++++++

Discord supports optional parameters, wherein a user doesn't need to provide a value during invocation.

A parameter is considered optional if its assigned a default value and/or annotated
to :obj:`~typing.Optional`.

For example, this command displays a given user's avatar, or the current user's avatar:

.. code-block:: python

    from typing import Optional

    @client.tree.command()
    async def avatar(interaction: discord.Interaction, user: Optional[discord.User] = None):
        avatar = (user or interaction.user).display_avatar
        await interaction.response.send_message(avatar.url)

On Discord:

.. image:: /images/guide/app_commands/avatar_command_optional_preview.png
    :width: 300

> `Python version 3.10+ union types <https://peps.python.org/pep-0604/>`_ are also supported instead of :obj:`typing.Optional`.

Reviewer 4: ```suggestion
:pep:`Python version 3.10+ union types <604>`_ are also supported instead of :obj:`typing.Optional`.
```
Again I wouldn't link to a PEP but this is slightly cleaner syntax, https://docs.python.org/3/library/stdtypes.html#types-union might be a better link.

    ):
        await interaction.response.send_message(
            f'i got: {mentionable}, of type: {mentionable.__class__.__name__}'
        )

Types that point to other types also don't have to include everything.
For example, a ``CHANNEL`` type parameter can point to any channel in a guild,
but can be narrowed down to a specific set of channels:

.. code-block:: python

    from typing import Union

    @client.tree.command()
    async def channel_info(interaction: discord.Interaction, channel: discord.abc.GuildChannel):
        # Everything except threads
        pass

    @client.tree.command()
    async def channel_info(interaction: discord.Interaction, channel: discord.TextChannel):
        # Only text channels
        pass

    @client.tree.command()
    async def channel_info(interaction: discord.Interaction, channel: Union[discord.Thread, discord.VoiceChannel]):
        # Threads and voice channels only
        pass

.. note::

    Union types can't mix Discord types.

    Something like ``Union[discord.Member, discord.TextChannel]`` isn't possible.

Refer to the :ref:`type conversion table <type_conversion>` for full information on sub-types.

Describing
+++++++++++

> Descriptions are added to parameters using the :func:`.app_commands.describe` decorator,

Reviewer 4: ```suggestion
Descriptions are added to parameters using the :func:`@app_commands.describe() <.app_commands.describe>` decorator,
```

where each keyword is treated as a parameter name.

.. code-block:: python

    @client.tree.command()
    @app_commands.describe(
        liquid='what type of liquid is on the wall',
        amount='how much of it is on the wall'
    )
    async def bottles(interaction: discord.Interaction, liquid: str, amount: int):
        await interaction.response.send_message(f'{amount} bottles of {liquid} on the wall!')

These show up on Discord just beside the parameter's name:

.. image:: /images/guide/app_commands/bottles_command_described.png

Not specifying a description results with an ellipsis "..." being used instead.

In addition to the decorator, parameter descriptions can also be added using
Google, Sphinx or NumPy style docstrings.

Examples using a command to add 2 numbers together:

.. tab:: NumPy

    .. code-block:: python

        @client.tree.command()
        async def addition(interaction: discord.Interaction, a: int, b: int):
            """adds 2 numbers together.

            Parameters
            -----------
            a: int
                left operand
            b: int
                right operand
            """

>             await interaction.response.send_message(f'{a} + {b} is {a + b}!')

Reviewer 4: ```suggestion
            await interaction.response.send_message(f'{a + b = }!')
```
Not too sure about whether or not you'd want to show this, up to you

.. tab:: Google

    .. code-block:: python

        @client.tree.command()
        async def addition(interaction: discord.Interaction, a: int, b: int):
            """adds 2 numbers together.

            Args:
                a (int): left operand
                b (int): right operand
            """

            await interaction.response.send_message(f'{a} + {b} is {a + b}!')

.. tab:: Sphinx

    .. code-block:: python

        @client.tree.command()
        async def addition(interaction: discord.Interaction, a: int, b: int):
            """adds 2 numbers together.

            :param a: left operand
            :param b: right operand
            """

            await interaction.response.send_message(f'{a} + {b} is {a + b}!')

Other meta info can be specified in the docstring, such as the function return type,
but in-practice only the parameter descriptions are used.

> Parameter descriptions added using :func:`.app_commands.describe` always

Reviewer 4: ```suggestion
Parameter descriptions added using :func:`@app_commands.describe() <.app_commands.describe>` always
```

take precedence over ones specified in the docstring.

Naming
^^^^^^^

Since parameter names are confined to the rules of Python's syntax,

> the library offers a method to rename them with the :func:`.app_commands.rename` decorator.

Reviewer 4: ```suggestion
the library offers a method to rename them with the :func:`@app_commands.rename() <.app_commands.rename>` decorator.
```

In use:

.. code-block:: python

    @client.tree.command()
    @app_commands.rename(amount='liquid-count')
    async def bottles(interaction: discord.Interaction, liquid: str, amount: int):
        await interaction.response.send_message(f'{amount} bottles of {liquid} on the wall!')

When referring to a renamed parameter in other decorators, the original parameter name should be used.
For example, to use :func:`~.app_commands.describe` and :func:`~.app_commands.rename` together:

.. code-block:: python

    @client.tree.command()
    @app_commands.rename(amount='liquid-count')
    @app_commands.describe(
        liquid='what type of liquid is on the wall',
        amount='how much of it is on the wall'
    )
    async def bottles(interaction: discord.Interaction, liquid: str, amount: int):
        await interaction.response.send_message(f'{amount} bottles of {liquid} on the wall!')

Choices
++++++++

:class:`str`, :class:`int` and :class:`float` type parameters can optionally set a list of choices for an argument

> using the :func:`.app_commands.choices` decorator.

Reviewer 4: ```suggestion
using the :func:`app_commands.choices() <.app_commands.choices>` decorator.
```

        Choice(name='Red', value=0xFF0000),
        Choice(name='Green', value=0x00FF00),
        Choice(name='Blue', value=0x0000FF)
    ])
    async def colour(interaction: discord.Interaction, colour: Choice[int]):
        """show a colour"""

        embed = discord.Embed(title=colour.name, colour=colour.value)
        await interaction.response.send_message(embed=embed)

On the client:

.. image:: /images/guide/app_commands/colour_command_preview.png
    :width: 400

discord.py also supports 2 other pythonic ways of adding choices to a command,
shown :func:`here <discord.app_commands.choices>` in the reference.

.. _autocompletion:

Autocompletion
+++++++++++++++

Autocomplete callbacks allow the bot to dynamically return up to 25 choices
to a user as they type a parameter.

In short:

- User starts typing.

- After a brief debounced pause from typing, Discord requests a list of choices from the bot.

- An autocomplete callback is called with the current user input.

- Returned choices are sent back to Discord and shown in the user's client.

  - An empty list can be returned to denote no choices.

Attaching an autocomplete function to a parameter can be done in 2 main ways:

> 1. From the command, with the :meth:`~.app_commands.Command.autocomplete` decorator
> 2. With a separate decorator, :func:`.app_commands.autocomplete`

Reviewer 4: ```suggestion
1. From the command, with the :meth:`@Command.autocomplete() <.app_commands.Command.autocomplete>` decorator
2. With a separate decorator, :func:`@app_commands.autocomplete() <.app_commands.autocomplete>`
```
(not sure if the class name should show up here)

Code examples for either method can be found in the corresponding reference page.

.. note::

>     Unlike :func:`.app_commands.choices`, a user can still submit any value instead of

Reviewer 4: ```suggestion
    Unlike :func:`@app_commands.choices <.app_commands.choices>`, a user can still submit any value instead of
```

If you're familar with the commands extension (:ref:`ext.commands <discord_ext_commands>`), a lot of similarities can be drawn between transformers and converters.

To use this transformer in a command, a paramater needs to annotate to :class:`.app_commands.Transform`,
passing the transformed type and transformer respectively.

.. code-block:: python

    from discord.app_commands import Transform

    @client.tree.command()
    async def date(interaction: discord.Interaction, when: Transform[datetime.datetime, DateTransformer]):
        # do something with 'when'...

It's also possible to instead pass an instance of the transformer instead of the class directly,
which opens up the possibility of setting up some state in :meth:`~object.__init__`.

Since the parameter's type annotation is replaced with :class:`~.app_commands.Transform`,
the underlying type and other information must now be provided through the :class:`~.app_commands.Transformer` itself.

These can be provided by overriding the following properties:

- :attr:`~.Transformer.type`
- :attr:`~.Transformer.min_value`
- :attr:`~.Transformer.max_value`
- :attr:`~.Transformer.choices`
- :attr:`~.Transformer.channel_types`

> Since these are properties, they must be decorated with :class:`property`:

Reviewer 4: ```suggestion
Since these are properties, they must be decorated with :class:`@property <property>`:
```

    @client.tree.command()
    async def userinfo(
        interaction: discord.Interaction,
        user: Union[discord.User, discord.Member]
    ):
        info = user.name

        # add some extra info if this command was invoked in a guild
        if isinstance(user, discord.Member):
            joined = user.joined_at
            if joined:
                relative = discord.utils.format_dt(joined, 'R')
                info = f'{info} (joined this server {relative})'

        await interaction.response.send_message(info)

.. _command_groups:

Command groups
---------------

To make a more organised and complex tree of commands, Discord implements command groups and subcommands.
A group can contain up to 25 subcommands or subgroups, with up to 1 level of nesting supported.

Meaning, a structure like this is possible:

.. code-block::

    todo
    ├── lists
    │   ├── /todo lists create
    │   └── /todo lists switch
    ├── /todo add
    └── /todo delete

Command groups **are not invocable** on their own.

Therefore, instead of creating a command the standard way by decorating an async function,
groups are created by using :class:`.app_commands.Group`.

> This class is customisable by subclassing and passing in any relevant fields at inheritance:

Reviewer 4: ```suggestion
This class can be subclassed and attributes passed to the class constructor:
```
IIRC this is the "class constructor" but I can't find anything linking to it 😎

.. code-block:: python

    class Todo(app_commands.Group, description='manages a todolist'):
        ...

    client.tree.add_command(Todo()) # required!

.. note::

    Groups need to be added to the command tree manually with :meth:`.CommandTree.add_command`,

>     since we lose the shortcut decorator :meth:`.CommandTree.command` with this class approach.

Reviewer 4: ```suggestion
    since we lose the shortcut decorator :meth:`@CommandTree.command() <.CommandTree.command>` with this class approach.
```

            description='commands for managing different todolists for different purposes'
        )

        @todo_lists.command(name='switch', description='switch to a different todolist')
        async def todo_lists_switch(self, interaction: discord.Interaction):
            ... # /todo lists switch

.. image:: /images/guide/app_commands/todo_group_nested_preview.png
    :width: 400

Nested group commands can be moved into another class if it ends up being a bit too much to read in one class:

.. code-block:: python

    class TodoLists(app_commands.Group, name='lists'):
        """commands for managing different todolists for different purposes"""

        @app_commands.command(name='switch', description='switch to a different todolist')
        async def todo_lists_switch(self, interaction: discord.Interaction):
            ...

    class Todo(app_commands.Group, description='manages a todolist'):
        @app_commands.command(name='add', description='add a todo')
        async def todo_add(self, interaction: discord.Interaction):
            await interaction.response.send_message('added something to your todolist...!')

        todo_lists = TodoLists()

> Decorators like :func:`.app_commands.default_permissions` and :func:`.app_commands.guild_only`

Reviewer 4: ```suggestion
Decorators like :func:`@app_commands.default_permissions() <.app_commands.default_permissions>` and :func:`@app_commands.guild_only() <.app_commands.guild_only>`
```

can be added on top of a subclass to apply to the group, for example:

.. code-block:: python

    @app_commands.default_permissions(manage_emojis=True)
    class Emojis(app_commands.Group):
        ...

Due to a Discord limitation, individual subcommands cannot have differing official-checks.

.. _guild_commands:

Guild commands
---------------

So far, all the command examples in this page have been global commands,
which every guild your bot is in can see and use.

In contrast, guild commands are only seeable and usable by members of a certain guild.

There are 2 main ways to specify which guilds a command should sync a copy to:

> - Via the :func:`.app_commands.guilds` decorator, which takes a variadic amount of guilds

Reviewer 4: ```suggestion
- Via the :func:`@app_commands.guilds() <.app_commands.guilds>` decorator, which takes a variadic amount of guilds
```

    class MyClient(discord.Client):
        def __init__(self):
            super().__init__(intents=discord.Intents.default())
            self.tree = app_commands.CommandTree(self)

        async def setup_hook(self):
            guild = discord.Object(695868929154744360) # a bot testing server
            self.tree.copy_global_to(guild)
            await self.tree.sync(guild=guild)

You'll typically find this syncing paradigm in some of the examples in the repository.

.. _checks:

Checks
-------

Checks refer to the restrictions an app command can have for invocation.
A user needs to pass all checks on a command in order to be able to invoke and see the command on their client.

Age-restriction
++++++++++++++++

Indicates whether this command can only be used in NSFW channels or not.

This can be configured by passing the ``nsfw`` keyword argument within the command decorator:

.. code-block:: python

    @client.tree.command(nsfw=True)
    async def evil(interaction: discord.Interaction):
        await interaction.response.send_message('******') # very explicit text!

Guild-only
+++++++++++

Indicates whether this command can only be used in guilds or not.

> Enabled by adding the :func:`.app_commands.guild_only` decorator when defining an app command:

Reviewer 4: ```suggestion
Enabled by adding the :func:`@app_commands.guilds() <.app_commands.guild_only>` decorator when defining an app command:
```

.. code-block:: python

    @client.tree.command()
    @app_commands.guild_only()
    async def serverinfo(interaction: discord.Interaction):
        assert interaction.guild is not None
        await interaction.response.send_message(interaction.guild.name)

Default permissions
++++++++++++++++++++

This sets the default permissions a user needs in order to be able to see and invoke an app command.

> Configured by adding the :func:`.app_commands.default_permissions` decorator when defining an app command:

Reviewer 4: ```suggestion
Configured by adding the :func:`@app_commands.default_permissions() <.app_commands.default_permissions>` decorator when defining an app command:
```

.. code-block:: python

    @client.tree.command()
    @app_commands.default_permissions(manage_nicknames=True)
    async def nickname(interaction: discord.Interaction, newname: str):
        guild = interaction.guild
        if not guild:
            await interaction.response.send_message("i can't change my name here")
        else:
            await guild.me.edit(nick=newname)
            await interaction.response.send_message(f'hello i am {newname} now')

Commands with this check are still visible and invocable in the bot's direct messages,
regardless of the permissions specified.

> To prevent this, :func:`~.app_commands.guild_only` can also be added.

Reviewer 4: ```suggestion
To prevent this, :func:`@guild_only() <.app_commands.guild_only>` can also be added.
```

        await interaction.response.send_message("you're lucky!")

Transforming the check into its own decorator for easier usage:

.. code-block:: python

    import random

    def coinflip():
        async def predicate(interaction: discord.Interaction) -> bool:
            return random.randint(0, 1) == 1
        return app_commands.check(predicate)

    @client.tree.command()
    @coinflip()
    async def fiftyfifty(interaction: discord.Interaction):
        await interaction.response.send_message("you're lucky!")

Checks are called sequentially and retain decorator order, bottom-to-top.

Take advantage of this order if, for example, you only want a cooldown to apply if a previous check passes:

.. code-block:: python

    @client.tree.command()
    @app_commands.checks.cooldown(1, 5.0) # called second
    @coinflip() # called first
    async def fiftyfifty(interaction: discord.Interaction):
        await interaction.response.send_message("you're very patient and lucky!")

Custom checks can either be:

- local, only running for a single command (as seen above).

- on a group, running for all child commands, and before any local checks.

>  - Added using the :meth:`.app_commands.Group.error` decorator or overriding :meth:`.app_commands.Group.on_error`.

Reviewer 4: ```suggestion
 - Added using the :meth:`@app_commands.Group.error() <.app_commands.Group.error>` decorator or overriding :meth:`.app_commands.Group.on_error`.
```

.. _global_check:

Global check
+++++++++++++

To define a global check, override :meth:`.CommandTree.interaction_check` in a :class:`~.app_commands.CommandTree` subclass.
This method is called before every command invoke.

For example:

.. code-block:: python

    whitelist = {
        # cool people only
        236802254298939392,
        402159684724719617,
        155863164544614402
    }

    class CoolPeopleTree(app_commands.CommandTree):
        async def interaction_check(self, interaction: discord.Interaction) -> bool:
            return interaction.user.id in whitelist

.. note::

    If your project uses :class:`.ext.commands.Bot` as the client instance,
    the :class:`.CommandTree` class can be configured via
    the ``tree_cls`` keyword argument in the bot constructor:

    .. code-block:: python

        from discord.ext import commands

        bot = commands.Bot(
            command_prefix='?',
            intents=discord.Intents.default(),
            tree_cls=CoolPeopleTree
        )

.. _error_handling:

Error handling
---------------

So far, any exceptions raised within a command callback, any custom checks, in a transformer

> or during localisation, et cetera should just be logged in the program's ``stderr`` or through any custom logging handlers.

Reviewer 4: ```suggestion
or during localisation, et cetera should just be logged in the program's :obj:`sys.stderr` or through any custom logging handlers.
```

In order to catch exceptions and do something else, such as sending a message to let
a user know their invocation failed for some reason, the library uses something called error handlers.

There are 3 types of handlers:

1. A local handler, which only catches exceptions for a specific command

>    Attached using the :meth:`.app_commands.Command.error` decorator.

Reviewer 4: ```suggestion
   Attached using the :meth:`@app_commands.Command.error() <.app_commands.Command.error>` decorator.
```

2. A group handler, which catches exceptions only for a certain group's subcommands.

>    Added by using the :meth:`.app_commands.Group.error` decorator or overriding :meth:`.app_commands.Group.on_error`.

Reviewer 4: ```suggestion
   Added by using the :meth:`@app_commands.Group.error() <.app_commands.Group.error>` decorator or overriding :meth:`.app_commands.Group.on_error`.
```

3. A global handler, which catches all exceptions in all commands.

>    Added by using the :meth:`.CommandTree.error` decorator or overriding :meth:`.CommandTree.on_error`.

Reviewer 4: ```suggestion
   Added by using the :meth:`@CommandTree.error() <.CommandTree.error>` decorator or overriding :meth:`.CommandTree.on_error`.
```

        if isinstance(error, Unlucky):
            await interaction.response.send_message(str(error))

Raising an exception from a transformer and catching it:

.. code-block:: python

    from discord.app_commands import Transform

    class BadDateArgument(app_commands.Translator):
        def __init__(self, argument: str):
            super().__init__(f'expected a date in dd/mm/yyyy format, not "{argument}".')

    class DateTransformer(app_commands.Transformer):
        async def transform(self, interaction: discord.Interaction, value: str) -> datetime.datetime:
            try:
                when = datetime.datetime.strptime(date, '%d/%m/%Y')
            except ValueError:
                raise BadDateArgument(value)

            when = when.replace(tzinfo=datetime.timezone.utc)
            return when

    # pretend `some_command` is a command that uses this transformer

    @some_command.error
    async def some_command_error(interaction: discord.Interaction, error: app_commands.AppCommandError):
        if isinstance(error, BadDateArgument):
            await interaction.response.send_message(str(error))

> Instead of printing plainly to :obj:`sys.stderr`, the standard ``logging`` module can be configured instead -

Reviewer 4: ```suggestion
Instead of printing plainly to :obj:`sys.stderr`, the standard :mod:`logging` module can be configured instead -
```

.. _translating:

Translating
------------

Discord supports localisation (l10n) for the following fields:

- Command names and descriptions
- Parameter names and descriptions
- Choice names (choices and autocomplete)

This allows the above fields to appear differently according to a user client's language setting.

Localisations can be done :ddocs:`partially <interactions/application-commands#localization>` -
when a locale doesn't have a translation for a given field, Discord will use the default/original string instead.

Support for l10n is implemented in discord.py with the :class:`.app_commands.Translator` interface,
which are effectively classes containing a core ``transform`` method that
takes the following parameters:

1. ``string`` - the string to be translated according to ``locale``
2. ``locale`` - the locale to translate to
3. ``context`` - the context of this translation (what type of string is being translated)

When :meth:`.CommandTree.sync` is called, this method is called in a heavy loop for each
string for each locale.

A wide variety of translation systems can be implemented using this interface, such as

> `gettext <https://docs.python.org/3/library/gettext.html>`_ and

Reviewer 4: ```suggestion
:mod:`gettext` and
```

`Project Fluent <https://projectfluent.org/>`_.

Only strings marked as ready for translation are passed to the method.
By default, every string is considered translatable and passed.

Nonetheless, to specify a translatable string explicitly,
simply pass a string wrapped in :class:`~.app_commands.locale_str` in places you'd usually use :class:`str`:

.. code-block:: python

    from discord.app_commands import locale_str as _

    @client.tree.command(name=_('example'), description=_('an example command'))
    async def example(interaction: discord.Interaction):
        ...

To toggle this behaviour, set the ``auto_locale_strings`` keyword-argument
to :obj:`False` when creating a command:

.. code-block:: python

    @client.tree.command(name='example', description='an example command', auto_locale_strings=False)
    async def example(interaction: discord.Interaction):

>         # i am not translated

Reviewer 4: ```suggestion
        ...  # i am not translated
```

+-----------------+------------------------------------------------------------------------------------+
|   Discord Type  |                                Python Type                                         |
+=================+====================================================================================+
| ``STRING``      | :class:`str`                                                                       |
+-----------------+------------------------------------------------------------------------------------+
| ``INTEGER``     | :class:`int`                                                                       |
+-----------------+------------------------------------------------------------------------------------+
| ``BOOLEAN``     | :class:`bool`                                                                      |
+-----------------+------------------------------------------------------------------------------------+
| ``NUMBER``      | :class:`float`                                                                     |
+-----------------+------------------------------------------------------------------------------------+

> | ``USER``        | :class:`~discord.User` or :class:`~discord.Member`                                 |

Reviewer 3: I feel like it's worth noting that when using `discord.Member` as a typehint, that it will not have any presence data (e.g. there is no data on the online status of an member)

Reviewer 3: Thinking about it, you are right about it

Preferably, this tree should be attached to the client instance to
play well with type checkers and to allow for easy access from anywhere in the code:

.. code-block:: python3

    class MyClient(discord.Client):
        def __init__(self):
            super().__init__(intents=discord.Intents.default())
            self.tree = app_commands.CommandTree(self)

    client = MyClient()

.. note::

    If your project instead uses :class:`.ext.commands.Bot` as the client instance,
    a :class:`~discord.app_commands.CommandTree` has already been defined at :attr:`.Bot.tree`,
    so this step is largely skipped.

Creating a command
-------------------

Slash commands are created by decorating an async function.
This function is then called whenever the slash command is invoked.

For example, the following code responds with "meow" on invocation:

.. code-block:: python3

    @client.tree.command()
    async def meow(interaction: discord.Interaction):
        """Meow meow meow"""

        await interaction.response.send_message("meow")

Functions of this pattern are called callbacks, since their execution is
relinquished to the library to be called later.

> Callbacks always have one parameter, ``interaction``,

Reviewer 2: I think this wording "always have one parameter" might be confuse some people into thinking callbacks can have only a single parameter and nothing more, although additional parameters are covered later in this guide as well. Maybe "always have at least one parameter" instead?

.. _discord_slash_commands:

Slash commands
===============

Slash commands are one of Discord's primary methods of implementing a user-interface for bots.
Analogous to their name, they're previewed and invoked in the Discord client by beginning your message with a forward-slash.

.. image:: /images/guide/app_commands/meow_command_preview.png
    :width: 400

App commands are implemented within the :ref:`discord.app_commands <discord_app_commands>` package.
Code examples in this page will always assume the following two imports:

.. code-block:: python3

    import discord
    from discord import app_commands

Setting up
-----------

To work with app commands, bots need the ``applications.commands`` scope.

You can enable this scope when generating an OAuth2 URL for your bot, shown :ref:`here <discord_invite_bot>`.

Defining a Tree
++++++++++++++++

> First, a :class:`.app_commands.CommandTree` needs to be instaniated

Reviewer 5: ```suggestion
First, a :class:`.app_commands.CommandTree` needs to be instantiated
```
Fix typo.
